File: droneyee_ws12/droneyee_ws/src/rflysim_udp_picture_rec_topic_pub_rospkg/src/image_receive.cpp
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <fstream>
#include <dirent.h>
#include <iostream>
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <cv_bridge/cv_bridge.h>

using namespace std;
using namespace cv;

int PORT=9999;

int main(int argc, char* argv[])
{
    ros::init(argc, argv, "rec_node");
    ros::NodeHandle nh("~");
    image_transport::ImageTransport it(nh);
    string camera_id, port_str;
    string TYPE;
    string ENCODING; 
    string FRAME_ID;
    string TOPIC_NAME;
    nh.param<std::string>("camera_id", camera_id, "left");
    nh.param<int>("port", PORT, 9999);
    nh.param<std::string>("type", TYPE, "rgb");
    nh.param<std::string>("encoding", ENCODING, "bgr8");
    nh.param<std::string>("frame_id", FRAME_ID, "camera_color_optical_frame");
    nh.param<std::string>("topic_name", TOPIC_NAME, "/camera/rgb");
    cout << "camera_id: " << camera_id << endl;
    cout << "port: " << PORT << endl;
    //image_transport::Publisher pub = it.advertise("/camera/"+camera_id, 1);
    image_transport::Publisher pub = it.advertise(TOPIC_NAME, 1);

    struct sockaddr_in addr;
    socklen_t sockfd, addr_len = sizeof(struct sockaddr_in);

    /* 建立socket，注意必须是SOCK_DGRAM */
    if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
    {
        perror ("socket failed!");
        exit(1);
    }

    /* 填写sockaddr_in 结构 */
    bzero(&addr, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(PORT);
    addr.sin_addr.s_addr = htonl(INADDR_ANY) ;
 
    /* 绑定socket */
    if (bind(sockfd, (struct sockaddr *)&addr, sizeof(addr))<0)
    {
        perror("connect failed!");
        exit(1);
    }
 
    int imgFlag = 1234567890;
    //int fhead_size = 4*sizeof(int);
    int fhead_size = 24;
    //int img_head[4];
    int img_head[6];
    char data_total[900000];  //这个小了可能会出现段错误(核心已转储)的报错导致节点挂掉，所以这里由原先的600000改为900000
    Mat image;

    while (ros::ok) {
        int recvbytes = recvfrom(sockfd, (char*)img_head, fhead_size, 0, (struct sockaddr *)&addr, &addr_len);
        if (recvbytes != fhead_size || img_head[0] != imgFlag) {
            continue;
        }
        // cout << img_head[0] << "," << img_head[1] << "," << img_head[2] << "," << img_head[3] << endl << endl; 
        int imgLen = img_head[1];
        int imgPackUnit = img_head[2];
        int imgpackNum = img_head[3];
        int recvd_size = 0;

        for (int i=0; i<imgpackNum; i++) {
            int recvpack = recvfrom(sockfd, data_total+i*imgPackUnit, imgPackUnit, 0, (struct sockaddr *)&addr, &addr_len);
            recvd_size += recvpack;
        }

        if (recvd_size == imgLen) {
            // cout << "received image." << endl;
            //image = imdecode(Mat(1, recvd_size, CV_8UC1, data_total), IMREAD_COLOR);
            if(TYPE=="rgb")
            {
                image = imdecode(Mat(1, recvd_size, CV_8UC1, data_total), IMREAD_COLOR); 
            }
            if(TYPE=="gray")
            {
                image = imdecode(Mat(1, recvd_size, CV_8UC1, data_total), IMREAD_GRAYSCALE); 
            }
            if(TYPE=="depth")
            {
                image = imdecode(Mat(1, recvd_size, CV_8UC1, data_total), IMREAD_ANYDEPTH); 
            }
            //sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", image).toImageMsg();
            sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), ENCODING, image).toImageMsg();
            pub.publish(msg);
            // imshow("image", image);
            // if (char(waitKey(1))=='q') { break; }
        }
    }

    close(sockfd);
    return 0;
}

File: droneyee_ws12/droneyee_ws/src/rflysim_udp_picture_rec_topic_pub_rospkg/src/image_receive.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import socket  #导入socket模块
import time
import cv2
import numpy as np
import struct
import mmap
import json
import sys
import os
import rospy
import math
from cv_bridge import CvBridge
from sensor_msgs.msg import Image

      #server 接收端
      # 设置服务器默认端口号
#PORT = 9999
#PORT = 1000
      # 创建一个套接字socket对象，用于进行通讯
      # socket.AF_INET 指明使用INET地址集，进行网间通讯
      # socket.SOCK_DGRAM 指明使用数据协议，即使用传输层的udp协议



if __name__ == '__main__':
	rospy.init_node('image_recv')
        port = rospy.get_param('~port')
        IP = rospy.get_param('~IP')
        pub_topic_name = rospy.get_param('~pub_topic_name')
        TYPE = rospy.get_param('~type')
        FRAME_ID = rospy.get_param('~frame_id')
        ENCODING = rospy.get_param('~encoding')
	server_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
	#address = ("192.168.200.38", PORT) 
	#address = ("192.168.10.116", PORT) 
        #address = ("192.168.10.116", port)
        address = (IP, port)
	server_socket.bind(address)  # 为服务器绑定一个固定的地址，ip和端口
	#server_socket.settimeout(10)  #设置一个时间提示，如果10秒钟没接到数据进行提示


	imgFlag = 1234567890
	fhead_size = struct.calcsize('4i1d')
	StartFlag = False
	imgPackUnit = 60000
	recvd_size = 0
	data_total = b''

	ros_cv_bridge = CvBridge()

	#rgb_pub=rospy.Publisher("/rgb_image_topic",Image, queue_size =3)
        rgb_pub=rospy.Publisher(pub_topic_name,Image, queue_size =3)

	#rate = rospy.Rate(50)
	    
	while not rospy.is_shutdown():
	    buf,addr = server_socket.recvfrom(60000)  # 
	    #print('get message.\n')
	    #if len(buf)==fhead_size:
	    if len(buf)==24:
		#print('get head.\n')
		dd = struct.unpack('4i1d',buf)
		chksm = dd[0]
		if chksm==imgFlag:
		    #print('get right head.\n')
		    StartFlag=True
		    data_size = dd[1]
		    imgPackUnit= dd[2]
		    recvd_size = 0
		    data_total = b''
		    timeStmp = dd[4]
                    imgtimestampsec=math.floor(timeStmp)
                    imgtimestampnsec=(timeStmp-imgtimestampsec)*1000000000
		    continue

	    #后面相当于循环接收图片码流
	    if StartFlag:
		if data_size -recvd_size >imgPackUnit:
		    recvd_size += len(buf)
		else:
		    recvd_size = data_size  
		data_total += buf
		            
		if recvd_size == data_size:
		    nparr = np.frombuffer(data_total, np.uint8)
		    #colorType=cv2.IMREAD_COLOR
                    if TYPE == 'rgb':
                        colorType=cv2.IMREAD_COLOR
                    if TYPE == 'depth':
                        colorType=cv2.IMREAD_ANYDEPTH
                    if TYPE == 'gray':
                        colorType=cv2.IMREAD_GRAYSCALE
		    Img = cv2.imdecode(nparr, colorType)
		    print('True Data.\n')
                    #image_msg = ros_cv_bridge.cv2_to_imgmsg(Img, encoding="bgr8")
                    image_msg = ros_cv_bridge.cv2_to_imgmsg(Img, encoding=ENCODING)
                    #image_msg.header.frame_id = "camera_color_optical_frame"
                    image_msg.header.frame_id = FRAME_ID
                    image_msg.header.stamp.secs = imgtimestampsec
                    image_msg.header.stamp.nsecs = imgtimestampnsec
                    rgb_pub.publish(image_msg)
		    hasData = True
		    StartFlag=False
		    #cv2.imshow("caisetu",Img)
		    #cv2.waitKey(1) #1就能连续流畅弹出图片，0的话弹出一个得手动关掉程序才继续运行再弹出下一个。
		       
		    if recvd_size>data_size:
		        print('Wrong Data.')
		        StartFlag=False

	     
	    #rate.sleep()
     	#正常情况下接收数据并且显示，如果10秒钟没有接收数据进行提示（打印 "time out"）
	#当然可以不要这个提示，那样的话把"try:" 以及 "except"后的语句删掉就可以了


				      # 接收客户端传来的数据 recvfrom接收客户端的数据，默认是阻塞的，直到有客户端传来数据
				      # recvfrom 参数的意义，表示最大能接收多少数据，单位是字节
				      # recvfrom返回值说明
				      # receive_data表示接受到的传来的数据,是bytes类型
				      # client  表示传来数据的客户端的身份信息，客户端的ip和端口，元组
      #receive_data, client = server_socket.recvfrom(1024)
  

      #print("来自客户端%s,发送的%s\n" % (client, receive_data))  #打印接收的内容

